dlib/svm/pegasos.h

```diff
@@ -655,7 +655,7 @@ namespace dlib
 
         trainer_type trainer;
         scalar_type min_learning_rate;
-        bool verbose;
+        bool verbose = true;
```

Comment:
The first constructor doesn't initialize this member variable, which gcc (I think rightly) complained about now.

Can move the initialization to that constructor, if needed. Doing it like this is my personal preference, though.

Reply:
@arrufat
If you have time, feel free to check if the latest commit helped fix those `-Werror=maybe-uninitialized` issues on your machine, too?

Reply:
on it now :)

Reply:
When building from `dlib/tests` I get this:
```
[1/2] Building CXX object CMakeFiles/dtest.dir/trust_region.cpp.o
FAILED: CMakeFiles/dtest.dir/trust_region.cpp.o 
/usr/bin/c++  -I/home/adria/Projects/reunanen/dlib/.. -O3 -DNDEBUG   -W -Wall -Wextra -Werror -Wno-unused-function -Wno-strict-overflow -DDLIB_JPEG_SUPPORT -DDLIB_USE_BLAS -DDLIB_USE_LAPACK -DDLIB_USE_CUDA -DDLIB_PNG_SUPPORT -DDLIB_GIF_SUPPORT -Wreturn-type -MD -MT CMakeFiles/dtest.dir/trust_region.cpp.o -MF CMakeFiles/dtest.dir/trust_region.cpp.o.d -o CMakeFiles/dtest.dir/trust_region.cpp.o -c /home/adria/Projects/reunanen/dlib/test/trust_region.cpp
In file included from /home/adria/Projects/reunanen/dlib/test/trust_region.cpp:6:
In function ‘T dlib::test_functions::rosen(const dlib::matrix<T, 2, 1>&) [with T = double]’,
    inlined from ‘T {anonymous}::neg_rosen_model<T>::operator()({anonymous}::neg_rosen_model<T>::column_vector) const [with T = double]’ at /home/adria/Projects/reunanen/dlib/test/trust_region.cpp:37:44,
    inlined from ‘typename T::type dlib::negate_tr_model<funct_model>::operator()(const T&) const [with T = dlib::matrix<double, 0, 1>; funct_model = {anonymous}::neg_rosen_model<double>]’ at /home/adria/Projects/reunanen/dlib/../dlib/optimization/optimization_trust_region.h:515:26,
    inlined from ‘double dlib::find_min_trust_region(stop_strategy_type, const funct_model&, typename funct_model::column_vector&, double) [with stop_strategy_type = dlib::objective_delta_stop_strategy; funct_model = dlib::negate_tr_model<{anonymous}::neg_rosen_model<double> >]’ at /home/adria/Projects/reunanen/dlib/../dlib/optimization/optimization_trust_region.h:423:29,
    inlined from ‘double dlib::find_max_trust_region(stop_strategy_type, const funct_model&, typename funct_model::column_vector&, double) [with stop_strategy_type = dlib::objective_delta_stop_strategy; funct_model = {anonymous}::neg_rosen_model<double>]’ at /home/adria/Projects/reunanen/dlib/../dlib/optimization/optimization_trust_region.h:553:38,
    inlined from ‘void {anonymous}::test_with_rosen() [with T = double]’ at /home/adria/Projects/reunanen/dlib/test/trust_region.cpp:72:36:
/home/adria/Projects/reunanen/dlib/test/optimization_test_functions.h:192:21: error: ‘*(const double*)((char*)&<unnamed> + offsetof(const dlib::matrix<double, 2, 1, dlib::memory_manager_stateless_kernel_1<char>, dlib::row_major_layout>,dlib::matrix<double, 2, 1, dlib::memory_manager_stateless_kernel_1<char>, dlib::row_major_layout>::<unnamed>))’ may be used uninitialized [-Werror=maybe-uninitialized]
  192 |             const T x = m(0);
      |                     ^
/home/adria/Projects/reunanen/dlib/test/trust_region.cpp: In function ‘void {anonymous}::test_with_rosen() [with T = double]’:
/home/adria/Projects/reunanen/dlib/test/trust_region.cpp:37:44: note: ‘<anonymous>’ declared here
   37 |             return -static_cast<T>(rosen<T>(x));
      |                                    ~~~~~~~~^~~
In file included from /home/adria/Projects/reunanen/dlib/test/trust_region.cpp:6:
In function ‘T dlib::test_functions::rosen(const dlib::matrix<T, 2, 1>&) [with T = double]’,
    inlined from ‘T {anonymous}::neg_rosen_model<T>::operator()({anonymous}::neg_rosen_model<T>::column_vector) const [with T = double]’ at /home/adria/Projects/reunanen/dlib/test/trust_region.cpp:37:44,
    inlined from ‘typename T::type dlib::negate_tr_model<funct_model>::operator()(const T&) const [with T = dlib::matrix<double, 0, 1>; funct_model = {anonymous}::neg_rosen_model<double>]’ at /home/adria/Projects/reunanen/dlib/../dlib/optimization/optimization_trust_region.h:515:26,
    inlined from ‘double dlib::find_min_trust_region(stop_strategy_type, const funct_model&, typename funct_model::column_vector&, double) [with stop_strategy_type = dlib::objective_delta_stop_strategy; funct_model = dlib::negate_tr_model<{anonymous}::neg_rosen_model<double> >]’ at /home/adria/Projects/reunanen/dlib/../dlib/optimization/optimization_trust_region.h:423:29,
    inlined from ‘double dlib::find_max_trust_region(stop_strategy_type, const funct_model&, typename funct_model::column_vector&, double) [with stop_strategy_type = dlib::objective_delta_stop_strategy; funct_model = {anonymous}::neg_rosen_model<double>]’ at /home/adria/Projects/reunanen/dlib/../dlib/optimization/optimization_trust_region.h:553:38,
    inlined from ‘void {anonymous}::test_with_rosen() [with T = double]’ at /home/adria/Projects/reunanen/dlib/test/trust_region.cpp:72:36:
/home/adria/Projects/reunanen/dlib/test/optimization_test_functions.h:193:21: error: ‘<unnamed>.dlib::matrix<double, 2, 1, dlib::memory_manager_stateless_kernel_1<char>, dlib::row_major_layout>::data.dlib::row_major_layout::layout<double, 2, 1, dlib::memory_manager_stateless_kernel_1<char>, 1>::data[1]’ may be used uninitialized [-Werror=maybe-uninitialized]
  193 |             const T y = m(1);
      |                     ^
/home/adria/Projects/reunanen/dlib/test/trust_region.cpp: In function ‘void {anonymous}::test_with_rosen() [with T = double]’:
/home/adria/Projects/reunanen/dlib/test/trust_region.cpp:37:44: note: ‘<anonymous>’ declared here
   37 |             return -static_cast<T>(rosen<T>(x));
      |                                    ~~~~~~~~^~~
cc1plus: all warnings being treated as errors
ninja: build stopped: subcommand failed.
```

Reply:
I don't get why GCC complains about `x` and `y` maybe not being initialized (at the line of initialization)

Reply:
I also get this error (which I am guessing @pfeatherstone might know what's going on):

```
[1/3] Building CXX object CMakeFiles/dtest.dir/serialize.cpp.o
FAILED: CMakeFiles/dtest.dir/serialize.cpp.o 
/usr/bin/c++  -I/home/adria/Projects/reunanen/dlib/.. -O3 -DNDEBUG   -W -Wall -Wextra -Werror -Wno-unused-function -Wno-strict-overflow -DDLIB_JPEG_SUPPORT -DDLIB_USE_BLAS -DDLIB_USE_LAPACK -DDLIB_USE_CUDA -DDLIB_PNG_SUPPORT -DDLIB_GIF_SUPPORT -Wreturn-type -MD -MT CMakeFiles/dtest.dir/serialize.cpp.o -MF CMakeFiles/dtest.dir/serialize.cpp.o.d -o CMakeFiles/dtest.dir/serialize.cpp.o -c /home/adria/Projects/reunanen/dlib/test/serialize.cpp
In file included from /home/adria/Projects/reunanen/dlib/../dlib/memory_manager/../binary_search_tree/../serialize.h:241,
                 from /home/adria/Projects/reunanen/dlib/../dlib/memory_manager/../binary_search_tree/binary_search_tree_kernel_2.h:11,
                 from /home/adria/Projects/reunanen/dlib/../dlib/memory_manager/memory_manager_kernel_3.h:11,
                 from /home/adria/Projects/reunanen/dlib/../dlib/memory_manager.h:8,
                 from /home/adria/Projects/reunanen/dlib/../dlib/conditioning_class.h:13,
                 from /home/adria/Projects/reunanen/dlib/../dlib/compress_stream.h:10,
                 from /home/adria/Projects/reunanen/dlib/test/serialize.cpp:8:
/usr/include/c++/11.1.0/variant: In instantiation of ‘std::enable_if_t<((__exactly_once<std::variant<_Types>::__accepted_type<_Tp&&, typename std::enable_if<__not_self<_Tp&&>, void>::type> > && is_constructible_v<std::variant<_Types>::__accepted_type<_Tp&&, typename std::enable_if<__not_self<_Tp&&>, void>::type>, _Tp>) && is_assignable_v<std::variant<_Types>::__accepted_type<_Tp&&, typename std::enable_if<__not_self<_Tp&&>, void>::type>&, _Tp>), std::variant<_Types>&> std::variant<_Types>::operator=(_Tp&&) [with _Tp = const char (&)[40]; _Types = {int, float, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, {anonymous}::immutable_type}; std::enable_if_t<((__exactly_once<std::variant<_Types>::__accepted_type<_Tp&&, typename std::enable_if<__not_self<_Tp&&>, void>::type> > && is_constructible_v<std::variant<_Types>::__accepted_type<_Tp&&, typename std::enable_if<__not_self<_Tp&&>, void>::type>, _Tp>) && is_assignable_v<std::variant<_Types>::__accepted_type<_Tp&&, typename std::enable_if<__not_self<_Tp&&>, void>::type>&, _Tp>), std::variant<_Types>&> = std::variant<int, float, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, {anonymous}::immutable_type>&; std::variant<_Types>::__accepted_type<_Tp&&> = std::variant<int, float, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, {anonymous}::immutable_type>::__accepted_type<const char (&)[40], void>; typename std::enable_if<__not_self<_Tp&&>, void>::type = void]’:
/home/adria/Projects/reunanen/dlib/test/serialize.cpp:1155:16:   required from here
/usr/include/c++/11.1.0/variant:1445:26: error: use of deleted function ‘std::variant<_Types>& std::variant<_Types>::operator=(std::variant<_Types>&&) [with _Types = {int, float, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, {anonymous}::immutable_type}]’
 1445 |                 operator=(variant(std::forward<_Tp>(__rhs)));
      |                 ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/c++/11.1.0/variant:1369:16: note: ‘std::variant<_Types>& std::variant<_Types>::operator=(std::variant<_Types>&&) [with _Types = {int, float, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, {anonymous}::immutable_type}]’ is implicitly deleted because the default definition would be ill-formed:
 1369 |       variant& operator=(variant&&) = default;
      |                ^~~~~~~~
/usr/include/c++/11.1.0/variant:1369:16: error: use of deleted function ‘std::_Enable_copy_move<false, false, false, false, _Tag>& std::_Enable_copy_move<false, false, false, false, _Tag>::operator=(std::_Enable_copy_move<false, false, false, false, _Tag>&&) [with _Tag = std::variant<int, float, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, {anonymous}::immutable_type>]’
In file included from /usr/include/c++/11.1.0/variant:38,
                 from /home/adria/Projects/reunanen/dlib/../dlib/memory_manager/../binary_search_tree/../serialize.h:241,
                 from /home/adria/Projects/reunanen/dlib/../dlib/memory_manager/../binary_search_tree/binary_search_tree_kernel_2.h:11,
                 from /home/adria/Projects/reunanen/dlib/../dlib/memory_manager/memory_manager_kernel_3.h:11,
                 from /home/adria/Projects/reunanen/dlib/../dlib/memory_manager.h:8,
                 from /home/adria/Projects/reunanen/dlib/../dlib/conditioning_class.h:13,
                 from /home/adria/Projects/reunanen/dlib/../dlib/compress_stream.h:10,
                 from /home/adria/Projects/reunanen/dlib/test/serialize.cpp:8:
/usr/include/c++/11.1.0/bits/enable_special_members.h:308:5: note: declared here
  308 |     operator=(_Enable_copy_move&&) noexcept                         = delete;
      |     ^~~~~~~~
In file included from /usr/include/c++/11.1.0/x86_64-pc-linux-gnu/bits/c++allocator.h:33,
                 from /usr/include/c++/11.1.0/bits/allocator.h:46,
                 from /usr/include/c++/11.1.0/string:41,
                 from /usr/include/c++/11.1.0/bits/locale_classes.h:40,
                 from /usr/include/c++/11.1.0/bits/ios_base.h:41,
                 from /usr/include/c++/11.1.0/ios:42,
                 from /usr/include/c++/11.1.0/ostream:38,
                 from /usr/include/c++/11.1.0/iostream:39,
                 from /home/adria/Projects/reunanen/dlib/test/serialize.cpp:5:
/usr/include/c++/11.1.0/ext/new_allocator.h: In instantiation of ‘void __gnu_cxx::new_allocator<_Tp>::construct(_Up*, _Args&& ...) [with _Up = {anonymous}::my_custom_type; _Args = {const {anonymous}::my_custom_type&}; _Tp = {anonymous}::my_custom_type]’:
/usr/include/c++/11.1.0/bits/alloc_traits.h:512:17:   required from ‘static void std::allocator_traits<std::allocator<_CharT> >::construct(std::allocator_traits<std::allocator<_CharT> >::allocator_type&, _Up*, _Args&& ...) [with _Up = {anonymous}::my_custom_type; _Args = {const {anonymous}::my_custom_type&}; _Tp = {anonymous}::my_custom_type; std::allocator_traits<std::allocator<_CharT> >::allocator_type = std::allocator<{anonymous}::my_custom_type>]’
/usr/include/c++/11.1.0/bits/stl_vector.h:1192:30:   required from ‘void std::vector<_Tp, _Alloc>::push_back(const value_type&) [with _Tp = {anonymous}::my_custom_type; _Alloc = std::allocator<{anonymous}::my_custom_type>; std::vector<_Tp, _Alloc>::value_type = {anonymous}::my_custom_type]’
/home/adria/Projects/reunanen/dlib/test/serialize.cpp:1163:23:   required from here
/usr/include/c++/11.1.0/ext/new_allocator.h:156:11: error: use of deleted function ‘{anonymous}::my_custom_type::my_custom_type(const {anonymous}::my_custom_type&)’
  156 |         { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
      |           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/adria/Projects/reunanen/dlib/test/serialize.cpp:430:12: note: ‘{anonymous}::my_custom_type::my_custom_type(const {anonymous}::my_custom_type&)’ is implicitly deleted because the default definition would be ill-formed:
  430 |     struct my_custom_type
      |            ^~~~~~~~~~~~~~
/home/adria/Projects/reunanen/dlib/test/serialize.cpp:430:12: error: use of deleted function ‘std::variant<_Types>::variant(const std::variant<_Types>&) [with _Types = {int, float, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, {anonymous}::immutable_type}]’
In file included from /home/adria/Projects/reunanen/dlib/../dlib/memory_manager/../binary_search_tree/../serialize.h:241,
                 from /home/adria/Projects/reunanen/dlib/../dlib/memory_manager/../binary_search_tree/binary_search_tree_kernel_2.h:11,
                 from /home/adria/Projects/reunanen/dlib/../dlib/memory_manager/memory_manager_kernel_3.h:11,
                 from /home/adria/Projects/reunanen/dlib/../dlib/memory_manager.h:8,
                 from /home/adria/Projects/reunanen/dlib/../dlib/conditioning_class.h:13,
                 from /home/adria/Projects/reunanen/dlib/../dlib/compress_stream.h:10,
                 from /home/adria/Projects/reunanen/dlib/test/serialize.cpp:8:
/usr/include/c++/11.1.0/variant:1366:7: note: ‘std::variant<_Types>::variant(const std::variant<_Types>&) [with _Types = {int, float, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, {anonymous}::immutable_type}]’ is implicitly deleted because the default definition would be ill-formed:
 1366 |       variant(const variant& __rhs) = default;
      |       ^~~~~~~
/usr/include/c++/11.1.0/variant:1366:7: error: use of deleted function ‘constexpr std::_Enable_copy_move<false, false, false, false, _Tag>::_Enable_copy_move(const std::_Enable_copy_move<false, false, false, false, _Tag>&) [with _Tag = std::variant<int, float, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, {anonymous}::immutable_type>]’
In file included from /usr/include/c++/11.1.0/variant:38,
                 from /home/adria/Projects/reunanen/dlib/../dlib/memory_manager/../binary_search_tree/../serialize.h:241,
                 from /home/adria/Projects/reunanen/dlib/../dlib/memory_manager/../binary_search_tree/binary_search_tree_kernel_2.h:11,
                 from /home/adria/Projects/reunanen/dlib/../dlib/memory_manager/memory_manager_kernel_3.h:11,
                 from /home/adria/Projects/reunanen/dlib/../dlib/memory_manager.h:8,
                 from /home/adria/Projects/reunanen/dlib/../dlib/conditioning_class.h:13,
                 from /home/adria/Projects/reunanen/dlib/../dlib/compress_stream.h:10,
                 from /home/adria/Projects/reunanen/dlib/test/serialize.cpp:8:
/usr/include/c++/11.1.0/bits/enable_special_members.h:303:15: note: declared here
  303 |     constexpr _Enable_copy_move(_Enable_copy_move const&) noexcept  = delete;
      |               ^~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/11.1.0/vector:66,
                 from /usr/include/c++/11.1.0/functional:62,
                 from /usr/include/c++/11.1.0/pstl/glue_algorithm_defs.h:13,
                 from /usr/include/c++/11.1.0/algorithm:74,
                 from /home/adria/Projects/reunanen/dlib/../dlib/compress_stream/../algs.h:110,
                 from /home/adria/Projects/reunanen/dlib/../dlib/compress_stream/compress_stream_kernel_1.h:6,
                 from /home/adria/Projects/reunanen/dlib/../dlib/compress_stream.h:6,
                 from /home/adria/Projects/reunanen/dlib/test/serialize.cpp:8:
/usr/include/c++/11.1.0/bits/stl_uninitialized.h: In instantiation of ‘_ForwardIterator std::uninitialized_copy(_InputIterator, _InputIterator, _ForwardIterator) [with _InputIterator = std::move_iterator<{anonymous}::my_custom_type*>; _ForwardIterator = {anonymous}::my_custom_type*]’:
/usr/include/c++/11.1.0/bits/stl_uninitialized.h:333:37:   required from ‘_ForwardIterator std::__uninitialized_copy_a(_InputIterator, _InputIterator, _ForwardIterator, std::allocator<_Tp>&) [with _InputIterator = std::move_iterator<{anonymous}::my_custom_type*>; _ForwardIterator = {anonymous}::my_custom_type*; _Tp = {anonymous}::my_custom_type]’
/usr/include/c++/11.1.0/bits/stl_uninitialized.h:355:2:   required from ‘_ForwardIterator std::__uninitialized_move_if_noexcept_a(_InputIterator, _InputIterator, _ForwardIterator, _Allocator&) [with _InputIterator = {anonymous}::my_custom_type*; _ForwardIterator = {anonymous}::my_custom_type*; _Allocator = std::allocator<{anonymous}::my_custom_type>]’
/usr/include/c++/11.1.0/bits/vector.tcc:474:3:   required from ‘void std::vector<_Tp, _Alloc>::_M_realloc_insert(std::vector<_Tp, _Alloc>::iterator, _Args&& ...) [with _Args = {const {anonymous}::my_custom_type&}; _Tp = {anonymous}::my_custom_type; _Alloc = std::allocator<{anonymous}::my_custom_type>; std::vector<_Tp, _Alloc>::iterator = std::vector<{anonymous}::my_custom_type>::iterator]’
/usr/include/c++/11.1.0/bits/stl_vector.h:1198:21:   required from ‘void std::vector<_Tp, _Alloc>::push_back(const value_type&) [with _Tp = {anonymous}::my_custom_type; _Alloc = std::allocator<{anonymous}::my_custom_type>; std::vector<_Tp, _Alloc>::value_type = {anonymous}::my_custom_type]’
/home/adria/Projects/reunanen/dlib/test/serialize.cpp:1163:23:   required from here
/usr/include/c++/11.1.0/bits/stl_uninitialized.h:138:72: error: static assertion failed: result type must be constructible from value type of input range
  138 |       static_assert(is_constructible<_ValueType2, decltype(*__first)>::value,
      |                                                                        ^~~~~
/usr/include/c++/11.1.0/bits/stl_uninitialized.h:138:72: note: ‘std::integral_constant<bool, false>::value’ evaluates to false
```

Reply:
Yes, I'm getting that last one (about `variant` in `test/serialize.cpp`), too. Was just about to write about it.

The first observation is that apparently we can make the error go away by removing the `,immutable_type` here on [this line](github.com/davisking/dlib/blob/7048f4c1d4b818c06a35da26527062e2ba8644d1/dlib/test/serialize.cpp#L451).

While just removing it doesn't quite sound like the right thing to do, looking at [this line](https://github.com/davisking/dlib/blob/7048f4c1d4b818c06a35da26527062e2ba8644d1/dlib/test/serialize.cpp#L1155) (from which the error is coming) makes me think that `q` doesn't really look like it's immutable to begin with? But didn't dig very deep into it just yet, so have probably missed something obvious.

Reply:
On the other hand, the error from `test/trust_region.cpp` I don't know how to reproduce :-/

Reply:
> On the other hand, the error from `test/trust_region.cpp` I don't know how to reproduce :-/

I am just building with GCC 11.1.0 from `dlib/tests` :/

Reply:
Yeah that warning about the trust region code is a GCC bug.  I've seen that in some versions of GCC as well.

So the thing to do here is probably to make warnings as errors off by default, and just configure the github actions tests to turn it on.

Reply:
Hmm, ok, but why `matrix<T,0,1>` [here](github.com/davisking/dlib/blob/7048f4c1d4b818c06a35da26527062e2ba8644d1/dlib/test/trust_region.cpp#L71)? 

Would it help if it said `matrix<T,2,1>` instead? Or maybe just `auto`?

Reply:
Actually, setting line 61 to `matrix<T, 0, 1> ans` makes it compile.

Reply:
Hmm, ok. Pushed that change.

Reply:
Yay, it builds for me!

Reply:
Definitely don't use `auto` with `matrix`. It will in general end up capturing an expression template that refers to temporary variables, which will then go out of scope instantly.  Which turns the expression object into something holding dangling references.